clippy_lints/src/methods.rs

```diff
@@ -1105,6 +1131,93 @@ fn lint_iter_cloned_collect(cx: &LateContext, expr: &hir::Expr, iter_args: &[hir
     }
 }
 
+fn lint_unnecessary_fold(cx: &LateContext, expr: &hir::Expr, fold_args: &[hir::Expr]) {
+    // Check that this is a call to Iterator::fold rather than just some function called fold
+    if !match_trait_method(cx, expr, &paths::ITERATOR) {
+        return;
+    }
+
+    assert!(fold_args.len() == 3,
+        "Expected fold_args to have three entries - the receiver, the initial value and the closure");
```

Comment:
Should this actually panic? I imagine a `debug!` and `return;` suffices

Reply:
Well, it's a logic bug, so I would have thought it should panic. I don't feel strongly though - I can change if it you'd prefer.